README.md

```diff
@@ -9,21 +9,72 @@ rustup target add riscv64gc-unknown-none-elf
 cargo build
 ```
 
-## running
+## Running
 
-There is a `Makefile` provided with targets for running various hosts in qemu.
+There is a `Makefile` provided with targets for running various hosts in QEMU.
 
 ### Prerequisites
 
-- rust toolchain from [rustup](rustup.rs)
-- rustup target add riscv64gc-unknown-none-elf
-- install `gcc-riscv64-unknown-elf` for `riscv64-unknown-elf-objcopy`
-- qemu-system-riscv64 - build using  qemu
-  [instructions](https://wiki.qemu.org/Hosts/Linux) with
+Toolchains:
+- Rust: Install [rustup](rustup.rs), then: `rustup target add riscv64gc-unknown-none-elf`
+- GCC: Install `gcc-riscv64-unknown-elf`
+
+QEMU:
+- Out-of-tree patches are required; see table below.
+- Build using QEMU [instructions](https://wiki.qemu.org/Hosts/Linux) with
   `--target-list=riscv64-softmmu`
-  - Note qemu 7.0.0 has a bug emulating riscv and a newer version from git must be used.
 
-## Test VM
+OpenSBI:
+- Use source from upstream: https://github.com/riscv-software-src/opensbi
+- Build: `PLATFORM=generic CROSS_COMPILE=riscv64-unknown-linux-gnu- make`
+
+Linux kernel (for booting Linux VMs):
+- Out-of-tree patches are required; see table below.
+- Build: `ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- make defconfig Image`
+
+Latest known-working branches:
+
+| Project | Branch |
+| ------- | ------ |
+| QEMU    | https://github.com/rivosinc/qemu/tree/salus-integration-08172022 |
+| Linux   | https://github.com/rivosinc/linux/tree/salus-integration-08172022 |
+| OpenSBI | use upstream: https://github.com/riscv-software-src/opensbi |
+
+### Linux VM
+
+The `make run_linux` target will boot a bare Linux kernel as the host VM,
+though it doesn't do much without any filesystems or emulated devices.
+
+To launch a more functional Debian VM:
+- Grab pre-built `riscv64-virt` images here: https://people.debian.org/~gio/dqib/
+- Launch QEMU with:
+  ```
+  qemu-system-riscv64 \
+   -M virt,aia=aplic-imsic,aia-guests=4 \
+   -cpu rv64,x-aia=true -m 2048 -smp "${NCPU}" \
+   -nographic \
+   -bios "${OPENSBI}" \
+   -kernel "${SALUS}" -device guest-loader,addr=0xc0200000,kernel="${KERNEL}" \
+   -append "console=hvc0 earlycon=sbi keep_bootcon root=LABEL=rootfs" \
+   -device guest-loader,addr=0xc2200000,initrd="${INITRD}" \
+   -device x-riscv-iommu-pci \
+   -drive file="${ROOTFS}",if=none,id=hd \
+   -device nvme,serial=deadbeef,drive=hd \
+   -netdev user,id=usernet,hostfwd=tcp:127.0.0.1:7722-0.0.0.0:22 \
+   -device e1000e,netdev=usernet
+  ```
+- SSH into the VM with `ssh root@localhost -p 7722`
+
+Only PCI devices using MSI/MSI-X will be usable by the Linux VM. `virtio-pci`
+devices may also be used with `iommu_platform=on,disable-legacy=on` flags.
+For example:
+
+```
+-fsdev local,id=workspace,path=/workspace,security_model=none \
+-device virtio-9p-pci,fsdev=workspace,mount_tag=lworkspace,iommu_platform=on,disable-legacy=on \
+```
+
```

Comment:
Not sure if it's just me, but the terminal where the qemu is running did not go into the console prompt (#).
```
[  OK  ] Reached target Graphical Interface.
         Starting Record Runlevel Change in UTMP...
[   39.883908] e1000e 0000:00:03.0 eth0: NIC Link is Up 1000 Mbps Full Duplex, Flow Control: Rx/Tx
[   39.883908] e1000e 0000:00:03.0 eth0: NIC Link is Up 1000 Mbps Full Duplex, Flow Control: Rx/Tx
[   39.894644] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   39.894644] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[  OK  ] Finished Record Runlevel Change in UTMP.
```
It actually booted completely as I was able to login via ssh.

Reply:
Expected since we don't support the legacy `sbi_getchar()` extension.